backend/selectgen.ml

```diff
@@ -1308,49 +1389,62 @@ method emit_stores env data regs_addr =
 method private insert_return (env:environment) r (traps:trap_action list) =
   match r with
     None -> ()
-  | Some r ->
+  | Some (r, unclosed_regions) ->
+      self#insert_endregions env unclosed_regions;
       let loc = Proc.loc_results (Reg.typv r) in
-      if env.region_tail then
-        self#insert env (Iop Iendregion) (List.hd env.regions) [||];
       self#insert_moves env r loc;
       self#insert env (Ireturn traps) loc [||]
 
 method private emit_return (env:environment) exp traps =
-  self#insert_return env (self#emit_expr env exp) traps
-
+  self#insert_return env (self#emit_expr_aux env exp) traps
+
+method private tail_call_possible (env:environment) (pos:Lambda.region_close) =
+  match env.regions, pos with
+  | [], Rc_close_at_apply ->
+     Misc.fatal_error "Selectgen: Rc_close_at_apply with no region to close"
+  | [ _ ], Rc_close_at_apply -> true
+  | [], _ -> true
+
+  | _ :: _ :: _, Rc_close_at_apply
+  | _ :: _, (Rc_normal | Rc_nontail) ->
+     (* Some regions need to be closed after this call *)
+     false
+
+(* Emit an expression in tail position of a function,
+   closing all regions in [env.regions] *)
 method emit_tail (env:environment) exp =
-  let env' =
-    if env.region_tail then {env with region_tail=false} else env in
   match exp with
     Clet(v, e1, e2) ->
-      begin match self#emit_expr env' e1 with
+      begin match self#emit_expr env e1 with
         None -> ()
       | Some r1 -> self#emit_tail (self#bind_let env v r1) e2
       end
   | Clet_mut (v, k, e1, e2) ->
-     begin match self#emit_expr env' e1 with
+     begin match self#emit_expr env e1 with
        None -> ()
      | Some r1 -> self#emit_tail (self#bind_let_mut env v k r1) e2
      end
   | Cphantom_let (_var, _defining_expr, body) ->
       self#emit_tail env body
   | Cop((Capply(ty, ((Rc_close_at_apply | Rc_normal) as pos))) as op,
-        args, dbg) ->
-      let tail = (pos = Lambda.Rc_close_at_apply) in
-      let endregion = env.region_tail in
-      begin match self#emit_parts_list env' args with
+        args, dbg)
+       when self#tail_call_possible env pos ->
+      let close_region =
+        match pos, env.regions with
+        | Rc_close_at_apply, [r] -> Some r
```

Comment:
Maybe just use `List.hd` here since it is equivalent and makes clear that you are not making a decision at this point in the code based on the number of regions. If you do that you should expand the `_` pattern too.

Reply:
I added a case:
```
| Rc_close_at_apply, _ -> assert false
```
(I find this clearer than List.hd since List.hd suggests that the list also has a tl, and in this case we know the list to be a singleton)

Reply:
That seems fine, although I'd slightly prefer pushing the match on `env.regions` into the branch to make it clearer which data is being used to make a decision.